pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -403,6 +403,10 @@ func (lb *ListenerBuilder) buildHTTPConnectionManager(httpOpts *httpListenerOpts
 	// TypedPerFilterConfig in route needs these filters.
 	filters = append(filters, xdsfilters.Fault, xdsfilters.Cors)
 	filters = append(filters, lb.push.Telemetry.HTTPFilters(lb.node, httpOpts.class)...)
+	// Add EmptySessionFilter so that it can be overridden at route level per service.
+	if httpOpts.class != istionetworking.ListenerClassSidecarInbound {
+		filters = append(filters, xdsfilters.EmptySessionFilter)
```

Comment:
We cannot enable experimental filter for all users

Reply:
This does not do any thing because there is no config. But do you want to add additional feature flag for it?

Reply:
Is EmtpySessionFilter compiled in envoy ? 

I don't think we need a feature flag if it is off by default and is enabled with an annotation - the annotation serves the same purpose as the flag. Question is if we have an annotation on the Gateway - so all listeners would have 
the ability to support sessions.

Reply:
> Is EmtpySessionFilter compiled in envoy

Yes. 

> Question is if we have an annotation on the Gateway - so all listeners would have
the ability to support sessions.

All listeners will have the ability to support sessions but it is decided based on service annotation while building route for that service.

Reply:
EmtpySessionFilter may not be compiled in all versions. We cannot add a hard dependency on an experimental envoy filter. Given its support status, it should be assumed the code does the worst possible thing - just because its *supposed* to do nothing doesn't mean it will.

Reply:
Ok. Added feature flag. PTAL

Reply:
Why does envoy require the empty one as a must

Reply:
https://github.com/envoyproxy/envoy/pull/20960 - See this. That is how it is implemented to support specific routes in Envoy

Reply:
Unfortunately that's what envoy does. However I have one concern here - not sure which versions of Istio envoy have this fix, and I believe Istio 1.17 still needs to be compatible with envoy 1.16 for safe upgrade.

Can you confirm:
- does Istio 1.16 has the PR ? 
- do we want to add an envoy version check as well ? If you see my top comment, I would make the persistent session flag an envoy metadata, so user can opt-in a specific gateway and we can also check envoy version ( in particular if we want to add header-based sessions - which are not compiled in 1.16 ).

An alternative that I tested and works well with 1.16 is to use an invalid path ( /does_not_exist_cookie_path ) which will skip the cookie generation on listener, and can be overriden per route.

Reply:
https://github.com/envoyproxy/envoy/pull/20960 - This fix is available in 1.16. We have used it via Envoy filter already